### PR TITLE
AP_Common: Add AP_FWVersionDefine to unify version definition between all vehicles

### PR DESCRIPTION
**Pull request by patrickelectric**
Remove unnecessary `version.cpp` definition for each vehicle with a common header.

### CODE REVIEW

**Comment by magicrub** [comment]
I'm definitely in favor of this, but need input from @rmackay9 and @tridge on how it may effect releases and why it wasn't already done when AP_Common/AP_FWVersionDefine.h was created in the first place.

**Comment by peterbarker** [comment]
@patrickelectric have you considered this header is included everywhere and now appears to include date-and-time information?

Putting that into the `.cpp` file means rather better caching....

It would be nice to share more code there.  It's possible that something in AP_Vehicle *might* work - but I'm not sure about how well the includes will work there.

**Comment by patrickelectric** [comment]
Hello @peterbarker.

> have you considered this header is included everywhere

How so ? `version.h` is only included once per vehicle.

```
grep "#include \"version.h\"" . -r --exclude-dir={modules,libraries}
./ArduPlane/Plane.cpp:#include "version.h"
./Tools/AP_Periph/version.cpp:#include "version.h"
./Tools/AP_Periph/version_definition.h:#include "version.h"
./AntennaTracker/Tracker.cpp:#include "version.h"
./ArduCopter/Copter.cpp:#include "version.h"
./Rover/Rover.cpp:#include "version.h"
./ArduSub/Sub.cpp:#include "version.h"
```
This will also be protected by the `FORCE_VERSION_H_INCLUDE`

> now appears to include date-and-time information

I did not change the firmware version definition, the date-and-time information was always there.



> Putting that into the `.cpp` file means rather better caching....

Indeed it could, but the file is only included once so I don't see how will this improve the compilation time. But I did a couple of tests here just to make sure, and as result I noticed no time difference, I got around 0.3s faster with this patch with 6 samples compared to master.


> It would be nice to share more code there. It's possible that something in AP_Vehicle _might_ work - but I'm not sure about how well the includes will work there.

I'm not sure how this will fix the 'header problem' (if it's a problem at all), but since each vehicle will need to have their own `version.h` file I don't see how will this be an alternative to the current patch.

**Comment by peterbarker** [comment]
@patrickelectric ... and that file is transitively included by everything - at least in the vehicle directory...

@lucasdemarchi you had a chunk of input when we moved to the current versioning files - in particular using the `.cpp` file.  Can you remember the specific reasons?

This is a fantastic cleanup so long as it doesn't screw up `ccache` or whatnot.

**Comment by lucasdemarchi** [comment]
@peterbarker  when this was done a long time ago, wherever the version definition was, it ended up included by a ton of other compilation units. So you have 1 header that is included by only one, and that gets included by several others.  It's particularly bad when the git commit hash is part of the version, because then it means that for each incremental build after a commit you are going to "build" several files.

If now the header is only included in one place, then it's probably fine not having the version.cpp. However, re-building version.cpp into version.o is probably faster than rebuilding, say, plane.cpp.

File: Rover/version.h [hunk 1]
```diff
@@ -15,3 +15,5 @@
 #define FW_MINOR 1
 #define FW_PATCH 0
 #define FW_TYPE FIRMWARE_VERSION_TYPE_DEV
+
+#include <AP_Common/AP_FWVersionDefine.h>
```
Comment (tridge) [inline Rover/version.h:19]: this brings in the structure declaration which seems wrong for the .h file

**Review by tridge — CHANGES_REQUESTED** [review]

File: Rover/version.h (hunk 1 above)
Comment (tridge) [reply Rover/version.h:19]: my concern is the fwver symbol being defined twice

**Comment by peterbarker** [comment]
Ah.  I misread this; it's included from the `.cpp` file, not the header file.

File: Rover/version.h (hunk 1 above)
Comment (lucasdemarchi) [reply Rover/version.h:19]: In theory that's what FORCE_VERSION_H_INCLUDE is protecting against. So however includes that file need to define FORCE_VERSION_H_INCLUDE to get the symbol defined. And in theory there is only one place in the vehicle code doing that

**Comment by patrickelectric** [comment]
@tridge I believe that was already answered.

File: libraries/AP_Common/AP_FWVersionDefine.h [hunk 2]
```diff
@@ -13,9 +13,9 @@
  * with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
-#define FORCE_VERSION_H_INCLUDE
-#include "version.h"
-#undef FORCE_VERSION_H_INCLUDE
+#ifndef FORCE_VERSION_H_INCLUDE
```
Comment (peterbarker) [inline libraries/AP_Common/AP_FWVersionDefine.h:16]:
```suggestion
// this include file *defines* the structure holding the version information for the ArduPilot binary.  It must only be 
// included in a single place, thus the following protection:

#ifndef FORCE_VERSION_H_INCLUDE
```

**Review by peterbarker — CHANGES_REQUESTED** [review]
@patrickelectric Yes, I think that question has been adequately answered - the structure is defined in a header file but that header file is included exactly once in a .cpp file - and that's guaranteed by a guard.

I think, however, since it was a bit of a surprise as to what was going on that a comment might be a good idea....

**Comment by patrickelectric** [comment]
@peterbarker done.

**Review by peterbarker — APPROVED** [review]
Really nice cleanup.

**Review by peterbarker — CHANGES_REQUESTED** [review]
Ah, whoops.  Just noticed that the first commit still has stuff in it from both AP_Common and AP_Periph.  These need to be split apart.

**Comment by patrickelectric** [comment]
@peterbarker done.

**Review by peterbarker — APPROVED** [review]

**Review by tridge — APPROVED** [review]

**Comment by tridge** [comment]
good to go in when CI passes

**Comment by peterbarker** [comment]
Merged, thanks!

**Pull request merged**